enclave-tls/src/core/etls_common.c

```diff
@@ -0,0 +1,199 @@
+#ifdef SGX
+		else if (!strcmp(log_level_str, "warn") || !strcmp(log_level_str, "WARN"))
+#else
+		else if (!strcasecmp(log_level_str, "warn"))
+#endif
+			return ENCLAVE_TLS_LOG_LEVEL_WARN;
+#ifdef SGX
+		else if (!strcmp(log_level_str, "error") || !strcmp(log_level_str, "ERROR"))
+#else
+		else if (!strcasecmp(log_level_str, "error"))
+#endif
+			return ENCLAVE_TLS_LOG_LEVEL_ERROR;
+#ifdef SGX
+		else if (!strcmp(log_level_str, "fatal") || !strcmp(log_level_str, "FATAL"))
+#else
+		else if (!strcasecmp(log_level_str, "fatal"))
+#endif
+			return ENCLAVE_TLS_LOG_LEVEL_FATAL;
+#ifdef SGX
+		else if (!strcmp(log_level_str, "off") || !strcmp(log_level_str, "OFF"))
+#else
+		else if (!strcasecmp(log_level_str, "off"))
+#endif
+			return ENCLAVE_TLS_LOG_LEVEL_NONE;
+	}
+
+	return ENCLAVE_TLS_LOG_LEVEL_DEFAULT;
+}
+
+enclave_tls_err_t etls_instance_init(const char *name,
+                                     __attribute__((unused)) const char *realpath,
+                                     __attribute__((unused)) void **handle
+                                     )
+{
+#ifdef SGX
+	if (!strcmp(name, "nullcrypto"))
+		libcrypto_wrapper_nullcrypto_init();
+	else if (!strcmp(name, "nullattester"))
+		libattester_null_init();
+	else if (!strcmp(name, "nullverifier"))
+		libverifier_null_init();
+	else if (!strcmp(name, "sgx_ecdsa")) {
+		libattester_sgx_ecdsa_init();
+		libverifier_sgx_ecdsa_init();
+        }
+	else if (!strcmp(name, "sgx_ecdsa_qve"))
+		libverifier_sgx_ecdsa_qve_init();
+	else if (!strcmp(name, "sgx_la")) {
+                libattester_sgx_la_init();
+		libverifier_sgx_la_init();
+        }
+	else if (!strcmp(name, "nulltls"))
+		libtls_wrapper_nulltls_init();
+	else if (!strcmp(name, "openssl")) {
+		libtls_wrapper_openssl_init();
+		libcrypto_wrapper_openssl_init();
+        } else
+		return ENCLAVE_TLS_ERR_NO_NAME;
+#else
+	*handle = dlopen(realpath, RTLD_LAZY);
+	if (*handle == NULL) {
+		ETLS_ERR("failed on dlopen(): %s\n", dlerror());
+		return -ENCLAVE_TLS_ERR_DLOPEN;
+	}
+#endif
+
+	return ENCLAVE_TLS_ERR_NONE;
+}
+
+ssize_t etls_write(int fd, const void *buf, size_t count)
+{
+	ssize_t rc;
+#ifdef SGX
+	int sgx_status = ocall_write(&rc, fd, buf, count);
+	if (SGX_SUCCESS != sgx_status) {
+		ETLS_ERR("sgx failed to write data, sgx status: 0x%04x\n", sgx_status);
+	}
+#else
+	rc = write(fd, buf, count);
+#endif
+
+	return rc;
+}
+
+ssize_t etls_read(int fd, void *buf, size_t count)
+{
+	ssize_t rc;
+#ifdef SGX
+	int sgx_status = ocall_read(&rc, fd, buf, count);
+	if (SGX_SUCCESS != sgx_status) {
+		ETLS_ERR("sgx failed to read data, sgx status: 0x%04x\n", sgx_status);
+	}
+#else
+	rc = read(fd, buf, count);
+#endif
+	return rc;
+}
+
+uint64_t etls_opendir(const char* name)
+{
+	uint64_t dir;
+#ifdef SGX
+	int sgx_status = ocall_opendir(&dir, name);
+	if (sgx_status != SGX_SUCCESS) {
+		ETLS_ERR("sgx failed to open %s, sgx status: 0x%04x\n", name, sgx_status);
+	}
+#else
+	dir = (uint64_t)opendir(name);
+#endif
+	return dir;
```

Comment:
Not sure why using uint64_t instead of `DIR *`? If SGX build doesn't support it, you can define it for SGX build, rather than using uint64_t to make SGX build working.

Or even define a etls_dir_t like etls_dirent.

Reply:
unify SGX and non-sgx interface